gravitee-apim-gateway/gravitee-apim-gateway-policy/src/main/java/io/gravitee/gateway/policy/impl/ConditionalReadWriteStream.java

```diff
@@ -0,0 +1,179 @@
+import io.gravitee.gateway.api.stream.WriteStream;
+import io.gravitee.gateway.core.condition.ConditionEvaluator;
+import io.gravitee.policy.api.PolicyChain;
+import io.gravitee.policy.api.PolicyResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConditionalReadWriteStream extends SimpleReadWriteStream<Buffer> {
+
+    private static final String GATEWAY_POLICY_INTERNAL_ERROR_KEY = "GATEWAY_POLICY_INTERNAL_ERROR";
+    public static final Logger LOGGER = LoggerFactory.getLogger(ConditionalReadWriteStream.class);
+
+    private final String policyId;
+    private final ReadWriteStream<Buffer> delegate;
+    private final String condition;
+    private final ExecutionContext context;
+    private final ConditionEvaluator<String> conditionEvaluator;
+    private final PolicyChain chain;
+
+    private Boolean isConditionTruthy;
+
+    public ConditionalReadWriteStream(
+        ReadWriteStream<Buffer> delegate,
+        PolicyChain chain,
+        ExecutionContext context,
+        String policyId,
+        String condition,
+        ConditionEvaluator<String> conditionEvaluator
+    ) {
+        this.policyId = policyId;
+        this.delegate = delegate;
+        this.condition = condition;
+        this.context = context;
+        this.conditionEvaluator = conditionEvaluator;
+        this.chain = chain;
+    }
+
+    /**
+     * Set body handler for both delegate and this instance
+     * @param handler will handle the buffer
+     * @return this instance with the good body handler
+     */
+    @Override
+    public SimpleReadWriteStream<Buffer> bodyHandler(Handler<Buffer> handler) {
+        delegate.bodyHandler(handler);
+        return super.bodyHandler(handler);
+    }
+
+    /**
+     * Set end handler for both delegate and this instance
+     * @param handler that will be called at the end
+     * @return this instance with the good end handler
+     */
+    @Override
+    public SimpleReadWriteStream<Buffer> endHandler(Handler<Void> handler) {
+        delegate.endHandler(handler);
+        return super.endHandler(handler);
+    }
+
+    /**
+     * Pause both delegate and this instance in case of full queue
+     * @return this instance of ReadStream<Buffer>
+     */
+    @Override
+    public ReadStream<Buffer> pause() {
+        delegate.pause();
+        return super.pause();
+    }
+
+    /**
+     * Resume both delegate and this instance in case of full queue
+     * @return this instance of ReadStream<Buffer>
+     */
+    @Override
+    public ReadStream<Buffer> resume() {
+        delegate.resume();
+        return super.resume();
+    }
+
+    /**
+     * When condition evaluate to true, then we write on the delegate (targeted policy), else, on this instance
+     * @param buffer is the one to write
+     * @return this instance to be able to continue the chaining
+     */
+    @Override
+    public SimpleReadWriteStream<Buffer> write(Buffer buffer) {
+        if (evaluateCondition()) {
```

Comment:
I would prefer to avoid evaluating the condition for each of the chunk. so you should keep track of the first invocation result, and reuse it for subsequent packets.

Reply:
evaluation is done only once, `isConditionTruthy` is null at the beginning and then set only the first time condition is evaluated.
then, we always return `isConditionTruthy`